### ISSUE TITLE
CA1859 does not apply to async methods returning a Task<T>

### ISSUE THREAD

**Issue by martincostello** [dotnet/roslyn-analyzers#7262]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859)

### Describe the improvement

The following method raises CA1859, suggesting that the return type is changed to `List<string>`:

```csharp
private static IList<string> GetList()
{
    return new List<string>();
}
```

The same pattern is however not detected for async methods returning a `Task<T>`. The following code does not suggest that `Task<IList<string>>` is changed to `Task<List<string>>`:

```csharp
private static async Task<IList<string>> GetListAsync()
{
    await Task.CompletedTask;
    return new List<string>();
}
```

### Describe suggestions on how to achieve the rule

Change the rule to consider the real type that is used to create the `Task<T>`.

### Additional context

None.
